File: src/main/java/com/example/aistore/service/TelemetryService.java
package com.example.aistore.service;

import com.example.aistore.dto.TelemetryRequest;
import com.example.aistore.entity.InteractionType;
import com.example.aistore.entity.SearchHistory;
import com.example.aistore.entity.User;
import com.example.aistore.entity.UserInteraction;
import com.example.aistore.repository.SearchHistoryRepository;
import com.example.aistore.repository.UserInteractionRepository;
import com.example.aistore.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * User preference profiler &amp; behavioral telemetry collector.
 */
@Service
public class TelemetryService {

    private static final Logger log = LoggerFactory.getLogger(TelemetryService.class);

    private final UserInteractionRepository interactionRepository;
    private final SearchHistoryRepository searchHistoryRepository;
    private final UserRepository userRepository;

    public TelemetryService(UserInteractionRepository interactionRepository,
                            SearchHistoryRepository searchHistoryRepository,
                            UserRepository userRepository) {
        this.interactionRepository = interactionRepository;
        this.searchHistoryRepository = searchHistoryRepository;
        this.userRepository = userRepository;
    }

    @Transactional
    public UserInteraction recordInteraction(Long userId, TelemetryRequest request) {
        User user = userId != null ? userRepository.findById(userId).orElse(null) : null;
        InteractionType type;
        try {
            type = InteractionType.valueOf(request.interactionType().toUpperCase());
        } catch (IllegalArgumentException e) {
            type = InteractionType.PRODUCT_VIEW;
        }
        UserInteraction interaction = new UserInteraction();
        interaction.setUser(user);
        interaction.setSessionId(request.sessionId());
        interaction.setInteractionType(type);
        interaction.setProductId(request.productId());
        interaction.setCategoryName(request.categoryName());
        interaction.setBrandName(request.brandName());
        interaction.setSearchQuery(request.searchQuery());
        interaction.setDurationSeconds(Math.max(0, request.durationSeconds()));
        return interactionRepository.save(interaction);
    }

    @Transactional
    public void recordSearch(User user, String sessionId, String query, int resultCount) {
        try {
            SearchHistory history = new SearchHistory();
            history.setUser(user);
            history.setSessionId(sessionId);
            history.setQuery(query == null ? "" : query.substring(0, Math.min(300, query.length())));
            history.setResultCount(resultCount);
            searchHistoryRepository.save(history);
        } catch (Exception e) {
            log.warn("Failed to record search history: {}", e.getMessage());
        }
    }
}

File: app/services/telemetry_service.py
"""Behavioral telemetry collector: clickstream events -> user interactions."""
from __future__ import annotations

import uuid

from sqlalchemy.orm import Session

from ..models import SearchHistory, User, UserInteraction


def generate_session_id() -> str:
    return uuid.uuid4().hex


class TelemetryService:
    def __init__(self, db: Session):
        self.db = db

    def record_interaction(
        self,
        interaction_type: str,
        user: User | None = None,
        session_id: str | None = None,
        product_id: int | None = None,
        category_name: str | None = None,
        brand_name: str | None = None,
        search_query: str | None = None,
        duration_seconds: int = 0,
    ) -> UserInteraction:
        interaction = UserInteraction(
            user_id=user.id if user else None,
            session_id=session_id or generate_session_id(),
            interaction_type=interaction_type,
            product_id=product_id,
            category_name=category_name,
            brand_name=brand_name,
            search_query=search_query,
            duration_seconds=duration_seconds,
        )
        self.db.add(interaction)
        if interaction_type == "SEARCH" and search_query:
            self.db.add(
                SearchHistory(
                    user_id=user.id if user else None,
                    session_id=session_id,
                    query=search_query,
                    result_count=0,
                )
            )
        self.db.flush()
        return interaction

    def record_search(self, query: str, result_count: int, user: User | None = None, session_id: str | None = None) -> None:
        self.db.add(
            SearchHistory(
                user_id=user.id if user else None,
                session_id=session_id,
                query=query,
                result_count=result_count,
            )
        )
        self.db.flush()